ui/src/app/components/extend-initial-credits-modal.tsx

```diff
@@ -0,0 +1,114 @@
+import * as React from 'react';
+import { useState } from 'react';
+
+import { Profile } from 'generated/fetch';
+
+import { Button, StyledExternalLink } from 'app/components/buttons';
+import {
+  Modal,
+  ModalBody,
+  ModalFooter,
+  ModalTitle,
+} from 'app/components/modals';
+import { Spinner } from 'app/components/spinners';
+import { profileApi } from 'app/services/swagger-fetch-clients';
+import colors from 'app/styles/colors';
+import { displayDateWithoutHours, minusDays } from 'app/utils/dates';
+import { notificationStore, serverConfigStore } from 'app/utils/stores';
+import { supportUrls } from 'app/utils/zendesk';
+
+import { AoU } from './text-wrappers';
+
+interface Props {
+  onClose: Function;
+  expirationDate: number;
+}
+
+const ExtensionDescription = ({ expirationDate }) => {
+  const {
+    initialCreditsValidityPeriodDays,
+    initialCreditsExpirationWarningDays,
+  } = serverConfigStore.get().config;
+  return (
+    <div>
+      Initial credits provided by the <AoU /> Research Program expire after{' '}
+      {initialCreditsValidityPeriodDays} days. You are eligible to request an
+      extension after{' '}
+      {displayDateWithoutHours(
+        minusDays(expirationDate, initialCreditsExpirationWarningDays)
+      )}{' '}
+      which is {initialCreditsExpirationWarningDays} days before your current
+      expiration date {displayDateWithoutHours(expirationDate)}.
+      <StyledExternalLink
+        href={supportUrls.initialCredits}
+        style={{ marginTop: '1rem' }}
+      >
+        Learn more about credits & setting up a billing account
+      </StyledExternalLink>
+    </div>
+  );
+};
+
+export const ExtendInitialCreditsModal = ({
+  onClose,
+  expirationDate,
+}: Props) => {
+  const [saving, setSaving] = useState(false); // saving refers to the loading request time
+
+  const onRequestExtension = () => {
+    let updatedProfile: Profile;
+    setSaving(true);
+    profileApi()
+      .extendInitialCreditExpiration()
+      .then((profileResponse) => {
+        updatedProfile = profileResponse;
+        notificationStore.set({
+          title: 'Initial credits extended',
+          message:
+            'Initial credits will now expire on ' +
+            displayDateWithoutHours(
+              profileResponse.initialCreditsExpirationEpochMillis
+            ),
+        });
+      })
+      .catch((error: Response) => {
+        if (typeof error.json === 'function') {
+          error.json().then((errorJson) =>
+            notificationStore.set({
+              title: 'Error extending Initial credits',
+              message: errorJson?.message,
+            })
+          );
+        } else {
+          notificationStore.set({
+            title: 'Error extending Initial credits',
+            message: error.toString(),
```

Comment:
IntelliJ says this will print something like `[object error]` - I think JSON.stringify(error) is what you want.